pegen/c_generator.py

```diff
@@ -479,6 +479,12 @@ def join_conditions(self, keyword: str, node: Any, names: List[str]) -> None:
 
     def emit_action(self, node: Alt) -> None:
         self.print(f"res = {node.action};")
+
+        self.print("if (res == NULL && PyErr_Occurred()) {")
```

Comment:
This means, we will have to refactor helper functions to not just return `NULL`, but also set an error, right?

Reply:
> This means, we will have to refactor helper functions to not just return `NULL`, but also set an error, right?

Yes as this is the only way to distinguish the functions that can return `NULL` to say that the parsing failed for that rule. Note that in many situations in which we are returning NULL are because some C-API call fail and those set the error indicator already. For the ones that not, we need to set some error or transform those into asserts.

Reply:
Another possibility is going back to not generate the check for the rules that can receive `NULL` as a valid result but I don't know if we can make that in a reliable/correct way...

Reply:
I guess, refactoring helper functions isn't going to be that difficult, so I suggest sticking up with this for now.

Reply:
I agree although this check *may* have some impact on performance, although maybe it does not matter much (my initial checks didn't raise anything notable, but it would be good if you could check as well meanwhile I do some more rigorous measurements).

Reply:
Nothing significant on my system as well. On average 3.430 seconds, with master being very very close to that.